multicore_runtime/cloud_logging_test.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2015 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import json
+import logging
+import os
+import unittest
+
+from cloud_logging import CloudLoggingHandler
+
+class CloudLoggingTestCase(unittest.TestCase):
+
+  EXPECTED_TRACE_ID = None
+  EXPECTED_OVERRIDDEN_TRACE_ID = '1234123412341234'
+  EXPECTED_MESSAGE = 'test message'
+
+  def setUp(self):
+    self.handler = CloudLoggingHandler()
+    self.record = logging.makeLogRecord({'msg': self.EXPECTED_MESSAGE,
+                                         'levelname': 'INFO'})
+    self.record_with_extra = logging.makeLogRecord(
+        {'msg': self.EXPECTED_MESSAGE,
+         'levelname': 'INFO',
+         'trace_id': self.EXPECTED_OVERRIDDEN_TRACE_ID,})
+
+  def test_file_name_is_correct(self):
+    self.assertTrue(self.handler.baseFilename.startswith('/var/log/app_engine/app.'))
+    self.assertTrue(self.handler.baseFilename.endswith('.json'))
+
+  def test_format(self):
+    msg = self.handler.format(self.record)
+    payload = json.loads(msg)
+    if self.EXPECTED_TRACE_ID:
+      self.assertEquals(payload['traceId'], self.EXPECTED_TRACE_ID)
+    else:
+      self.assertNotIn('traceId', payload)
+
+  def test_format_with_trace_id_as_extra(self):
+    msg = self.handler.format(self.record_with_extra)
+    payload = json.loads(msg)
+    self.assertEquals(payload['traceId'], self.EXPECTED_OVERRIDDEN_TRACE_ID)
+
+class CloudLoggingTestCaseWithTraceIdEnv(CloudLoggingTestCase):
```

Comment:
I need a separate class because the environment variable needs to be torn down even on test failure. The safe way to do this is to set it in setUp and remove it in tearDown, which is per-class.

An alternative would be to mock os.environ instead of actually setting the env variable, but that seemed more complex than necessary, and running the same tests under two different conditions via class inheritance is idiomatic for unittest.